whistl/datasets.py

```diff
@@ -0,0 +1,258 @@
+    subfiles = [os.path.join(data_root, f) for f in os.listdir(data_root)]
+    # Keep only data directories, not anything else that might be in data_root
+
+    data_dirs = []
+    for f in subfiles:
+        if ('SRP' in f or 'GSE' in f or 'E-MEXP' in f) and os.path.isdir(f):
+            data_dirs.append(f)
+
+    return data_dirs
+
+
+def extract_dirs_with_label(data_dirs, disease_label, sample_to_label):
+    ''' Split the list of directories passed in into those that contain samples with a given
+    disease, and those that don't
+
+    Arguments
+    ---------
+    data_dirs: list of str or Path
+        A list of directories containing gene expression data
+    disease_label: str
+        The name of the disease whose samples will be used in testing
+    sample_to_label: dict
+        A string to string dict mapping sample ids to their corresponding label string.
+        E.g. {'GSM297791': 'sepsis'}
+
+    Returns
+    -------
+    dirs_without_label: list of str or Path
+        The directories from data_dirs not containing any samples corresponding to disease_label
+    dirs_with_label: list of str or Path
+        The directories from data_dirs that do contain samples with the provided disease
+    '''
+    dirs_without_label = []
+    dirs_with_label = []
+
+    for data_dir in data_dirs:
+        study = os.path.basename(os.path.normpath(data_dir))
+        study_file_name = study + '.tsv'
+        data_file = os.path.join(data_dir, study_file_name)
+
+        sample_ids = None
+        with open(data_file, 'r') as in_file:
+            # The tsv header contains all the sample ids for the study
+            sample_ids = in_file.readline()
+            sample_ids = sample_ids.strip().split('\t')
+
+        for sample_id in sample_ids:
+            if sample_id in sample_to_label:
+                if sample_to_label[sample_id] == disease_label:
+                    dirs_with_label.append(data_dir)
+                    break
+        else:
+            # If the the dir isn't added to dirs_with_label, add to dirs_without_label
+            dirs_without_label.append(data_dir)
+
+    return dirs_without_label, dirs_with_label
+
+
+def get_gene_intersection(data_dirs):
+    '''Find the set of genes present in all samples in data_dirs
+
+    Arguments
+    ---------
+    data_dirs: list of str or Path
+        The path to the directories containing the studies used in model training
+
+    Returns
+    -------
+    genes_to_use: list of str
+        The list of gene ids for genes present in all samples in all data directories
+    '''
+    df_list = []
+
+    for data_dir in data_dirs:
+        study = os.path.basename(os.path.normpath(data_dir))
+        study_file_name = study + '.tsv'
+        data_file = os.path.join(data_dir, study_file_name)
+        curr_df = pd.read_csv(data_file, sep='\t')
+
+        curr_df = curr_df.set_index('Gene')
+        df_list.append(curr_df)
+
+    combined_df = pd.concat(df_list, axis=1, join='inner')
```

Comment:
I was doing that for a bit, the issue is that when I have multiple datasets containing subsets of the directories, they all need to have the same genes. Regardless, it won't matter soon because I'll switch over to the compendium dataset which has a fixed number of genes for all studies